src/d3a/models/market/two_sided_pay_as_bid.py

```diff
@@ -120,70 +171,44 @@ def accept_bid(self, bid: Bid, energy: float = None,
             f"trade rate: {trade_rate} market {market_bid.price / market_bid.energy}"
 
         orig_price = bid.original_bid_price if bid.original_bid_price is not None else bid.price
-        residual = None
+        residual_bid = None
 
         if energy <= 0:
             raise InvalidTrade("Energy cannot be negative or zero.")
         elif energy > market_bid.energy:
             raise InvalidTrade("Traded energy cannot be more than the bid energy.")
         elif energy < market_bid.energy:
-            # Partial bidding
-            energy_portion = energy / market_bid.energy
-            residual_price = (1 - energy_portion) * market_bid.price
-            residual_energy = market_bid.energy - energy
-
-            # Manually creating the bid in order to not double-charge the fee of the
-            # residual bid
-            changed_bid = Bid(
-                str(uuid.uuid4()), residual_price, residual_energy,
-                buyer, seller,
-                original_bid_price=(1 - energy_portion) * orig_price,
-                buyer_origin=market_bid.buyer_origin
-            )
-
-            self.bids[changed_bid.id] = changed_bid
-            self.bid_history.append(changed_bid)
-
-            self._notify_listeners(MarketEvent.BID_CHANGED,
-                                   existing_bid=bid, new_bid=changed_bid)
-            residual = changed_bid
-
-            revenue, fees, final_trade_rate = GridFees.calculate_trade_price_and_fees(
-                trade_offer_info, self.transfer_fee_ratio
-            )
-            self.market_fee += fees
-            final_price = energy * final_trade_rate
-            bid = Bid(bid.id, final_price, energy, buyer, seller,
-                      original_bid_price=energy_portion * orig_price,
-                      buyer_origin=bid.buyer_origin)
+            # partial bid trade
+            accepted_bid, residual_bid = self.split_bid(market_bid, energy, orig_price)
+            bid = accepted_bid
         else:
-            revenue, fees, final_trade_rate = GridFees.calculate_trade_price_and_fees(
-                trade_offer_info, self.transfer_fee_ratio
-            )
-            self.market_fee += fees
-            final_price = energy * final_trade_rate
-            bid = bid._replace(price=final_price)
+            # full bid trade, nothing further to do here
+            pass
+
+        trade_price = self.determine_bid_price(trade_offer_info, energy)
+        bid = bid._replace(price=trade_price)
 
         # Do not adapt grid fees when creating the bid_trade_info structure, to mimic
         # the behavior of the forwarded bids which use the source market fee.
         updated_bid_trade_info = GridFees.propagate_original_offer_info_on_bid_trade(
                           trade_offer_info, 0.0)
 
+        # Delete the accepted bid from self.bids:
+        self.bids.pop(bid.id, None)
```

Comment:
```suggestion
        try:
            self.bids.pop(bid.id)
        except KeyError:
            raise BidNotFoundException
```

Reply:
Actually, this is not needed at all, as it is done at the first line in accept_bid. Removing it. Thanks fort pointing it out!

Reply:
I was wrong. This is again the accepted bid. I will incorporate your proposal.